File: native/windows/tests/ScreenFix.Core.Tests/ConfigValidatorTests.cs
using ScreenFix.Core.Configuration;
using ScreenFix.Core.Geometry;

namespace ScreenFix.Core.Tests;

public sealed class ConfigValidatorTests
{
    [Fact]
    public void Validate_AcceptsPermanentDefaults()
    {
        var result = ConfigValidator.Validate(ValidConfiguration());

        Assert.True(result.IsValid);
        Assert.Null(result.Error);
    }

    [Fact]
    public void Validate_RejectsNullConfiguration()
    {
        AssertInvalid("configuration is required", null);
    }

    [Fact]
    public void Validate_RejectsUnknownSchema()
    {
        AssertInvalid(
            "schema version must be 1",
            ValidConfiguration() with { SchemaVersion = 2 });
    }

    [Fact]
    public void Validate_RejectsMissingDisplay()
    {
        AssertInvalid(
            "display is required",
            ValidConfiguration() with { Display = null! });
    }

    [Fact]
    public void Validate_RejectsEmptyStableId()
    {
        var config = ValidConfiguration();

        AssertInvalid(
            "display stable ID is required",
            config with { Display = config.Display with { StableId = " " } });
    }

    [Fact]
    public void Validate_RejectsEmptyDisplayName()
    {
        var config = ValidConfiguration();

        AssertInvalid(
            "display name is required",
            config with { Display = config.Display with { Name = string.Empty } });
    }

    [Fact]
    public void Validate_RejectsNonPositiveDisplayDimensions()
    {
        var config = ValidConfiguration();

        AssertInvalid(
            "display dimensions must be finite and positive",
            config with { Display = config.Display with { Width = 0 } });
        AssertInvalid(
            "display dimensions must be finite and positive",
            config with { Display = config.Display with { Height = -1 } });
    }

    [Fact]
    public void Validate_RejectsNonFiniteDisplayDimensions()
    {
        var config = ValidConfiguration();

        AssertInvalid(
            "display dimensions must be finite and positive",
            config with { Display = config.Display with { Width = double.NaN } });
        AssertInvalid(
            "display dimensions must be finite and positive",
            config with { Display = config.Display with { Height = double.PositiveInfinity } });
    }

    [Fact]
    public void Validate_RequiresExactlyThreeBands()
    {
        AssertInvalid(
            "exactly three bands are required",
            ValidConfiguration() with { Bands = [] });
        AssertInvalid(
            "exactly three bands are required",
            ValidConfiguration() with { Bands = null! });
    }

    [Fact]
    public void Validate_RejectsNonFiniteBandValues()
    {
        AssertInvalid(
            "band values must be finite",
            WithFirstBand(new RectD(double.NaN, 0, 0.2, 0.3)));
        AssertInvalid(
            "band values must be finite",
            WithFirstBand(new RectD(0.2, 0, double.NegativeInfinity, 0.3)));
    }

    [Fact]
    public void Validate_RejectsNonPositiveBandSize()
    {
        AssertInvalid(
            "band size must be positive",
            WithFirstBand(new RectD(0.2, 0, 0, 0.3)));
        AssertInvalid(
            "band size must be positive",
            WithFirstBand(new RectD(0.2, 0, 0.2, -0.1)));
    }

    [Fact]
    public void Validate_RejectsNegativeBandOrigin()
    {
        AssertInvalid(
            "band origin must be non-negative",
            WithFirstBand(new RectD(-0.01, 0, 0.2, 0.3)));
        AssertInvalid(
            "band origin must be non-negative",
            WithFirstBand(new RectD(0.2, -0.01, 0.2, 0.3)));
    }

    [Fact]
    public void Validate_RejectsBandBoundsAboveOne()
    {
        AssertInvalid(
            "band bounds must not exceed 1",
            WithFirstBand(new RectD(0.9, 0, 0.2, 0.3)));
        AssertInvalid(
            "band bounds must not exceed 1",
            WithFirstBand(new RectD(0.2, 0.9, 0.2, 0.2)));
    }

    private static ScreenFixConfig ValidConfiguration() => DefaultConfiguration.Create(
        new DisplayIdentity("display-1", "Ultrawide", 3440, 1440));

    private static ScreenFixConfig WithFirstBand(RectD band)
    {
        var config = ValidConfiguration();
        RectD[] bands = [band, config.Bands[1], config.Bands[2]];
        return config with { Bands = bands };
    }

    private static void AssertInvalid(string error, ScreenFixConfig? config)
    {
        var result = ConfigValidator.Validate(config);

        Assert.False(result.IsValid);
        Assert.Equal(error, result.Error);
    }
}

File: native/macos/Tests/ScreenFixTests/ConfigValidatorTests.swift
import ScreenFixCore

private let validDisplay = DisplayIdentity(
    stableId: "saved-uuid",
    name: "Ultrawide",
    width: 3440,
    height: 1440
)

private func configuration(
    schemaVersion: Int = 1,
    display: DisplayIdentity = validDisplay,
    bands: [NormalizedRect]? = nil
) -> ScreenFixConfiguration {
    ScreenFixConfiguration(
        schemaVersion: schemaVersion,
        enabled: true,
        display: display,
        bands: bands ?? DefaultConfiguration.make(for: validDisplay, enabled: true).bands
    )
}

private func expectConfigurationError(_ config: ScreenFixConfiguration) throws {
    do {
        try ConfigValidator.validate(config)
    } catch is ConfigurationError {
        return
    }
    throw TestFailure(description: "expected typed ConfigurationError")
}

let configValidatorTests = [
    TestCase(name: "ConfigValidator accepts the exact valid default") {
        try ConfigValidator.validate(configuration())
    },
    TestCase(name: "ConfigValidator rejects unknown schema versions") {
        try expectConfigurationError(configuration(schemaVersion: 0))
        try expectConfigurationError(configuration(schemaVersion: 2))
    },
    TestCase(name: "ConfigValidator rejects missing display identity") {
        for stableId in ["", "   \n"] {
            try expectConfigurationError(configuration(display: DisplayIdentity(
                stableId: stableId,
                name: "Ultrawide",
                width: 3440,
                height: 1440
            )))
        }
    },
    TestCase(name: "ConfigValidator rejects invalid display dimensions") {
        for value in [0.0, -1.0, Double.nan, Double.infinity, -Double.infinity] {
            try expectConfigurationError(configuration(display: DisplayIdentity(
                stableId: "saved-uuid",
                name: "Ultrawide",
                width: value,
                height: 1440
            )))
            try expectConfigurationError(configuration(display: DisplayIdentity(
                stableId: "saved-uuid",
                name: "Ultrawide",
                width: 3440,
                height: value
            )))
        }
    },
    TestCase(name: "ConfigValidator requires exactly three bands") {
        let validBands = DefaultConfiguration.make(for: validDisplay, enabled: true).bands
        try expectConfigurationError(configuration(bands: Array(validBands.prefix(2))))
        try expectConfigurationError(configuration(bands: validBands + [validBands[0]]))
    },
    TestCase(name: "ConfigValidator rejects non-finite band fields") {
        let invalids = [Double.nan, Double.infinity, -Double.infinity]
        for value in invalids {
            try expectConfigurationError(configuration(bands: [
                NormalizedRect(x: value, y: 0, w: 0.2, h: 0.2),
                NormalizedRect(x: 0, y: value, w: 0.2, h: 0.2),
                NormalizedRect(x: 0, y: 0, w: value, h: value),
            ]))
        }
    },
    TestCase(name: "ConfigValidator rejects out-of-range and non-positive bands") {
        let invalidBands = [
            NormalizedRect(x: -0.1, y: 0, w: 0.2, h: 0.2),
            NormalizedRect(x: 0, y: -0.1, w: 0.2, h: 0.2),
            NormalizedRect(x: 0, y: 0, w: 0, h: 0.2),
            NormalizedRect(x: 0, y: 0, w: 0.2, h: 0),
            NormalizedRect(x: 0.9, y: 0, w: 0.2, h: 0.2),
            NormalizedRect(x: 0, y: 0.9, w: 0.2, h: 0.2),
        ]
        let valid = NormalizedRect(x: 0, y: 0, w: 0.2, h: 0.2)
        for invalid in invalidBands {
            try expectConfigurationError(configuration(bands: [invalid, valid, valid]))
        }
    },
    TestCase(name: "ConfigValidator accepts bands touching right and bottom boundaries") {
        let bands = [
            NormalizedRect(x: 0.8, y: 0, w: 0.2, h: 1),
            NormalizedRect(x: 0, y: 0.8, w: 1, h: 0.2),
            NormalizedRect(x: 0, y: 0, w: 1, h: 1),
        ]
        try ConfigValidator.validate(configuration(bands: bands))
    },
]
